rust/libnewsboat/src/history.rs

```diff
@@ -0,0 +1,142 @@
+use std::fs::OpenOptions;
+use std::io::{self, BufRead, BufReader, BufWriter, Write};
+use std::path::Path;
+
+pub struct History {
+    idx: usize,
+    lines: Vec<String>,
+}
+
+impl History {
+    pub fn new() -> History {
+        History {
+            idx: 0,
+            lines: Vec::new(),
+        }
+    }
+    pub fn add_line(&mut self, line: String) {
+        if !line.is_empty() {
+            self.lines.insert(0, line)
+        }
+        self.idx = 0;
+    }
+    pub fn next(&mut self) -> String {
+        match self.idx {
+            0 => String::from(""),
+            _ => {
+                self.idx -= 1;
+                self.lines[self.idx].to_string()
+            }
+        }
+    }
+    pub fn prev(&mut self) -> String {
+        match self.lines.len() {
+            0 => String::from(""),
+            len if self.idx < len => {
+                let line = self.lines[self.idx].to_string();
+                self.idx += 1;
+                return line;
+            }
+            _ => self.lines[self.idx - 1].to_string(),
+        }
+    }
+    pub fn load_from_file<P: AsRef<Path>>(&mut self, path: P) -> io::Result<()> {
+        let file = OpenOptions::new().read(true).open(path)?;
+        let lines: Vec<String> = BufReader::new(file)
+            .lines()
+            .collect::<Result<Vec<_>, _>>()?;
+        lines.into_iter().for_each(|ln| self.add_line(ln));
+        Ok(())
+    }
+    pub fn save_to_file<P: AsRef<Path>>(&self, path: P, mut limit: usize) -> io::Result<()> {
+        if limit == 0 || self.lines.len() == 0 {
+            return Ok(());
+        }
+
+        let file = OpenOptions::new()
+            .write(true)
+            .create(true)
+            .truncate(true)
+            .open(path)?;
+        let mut f = BufWriter::new(file);
+
+        if limit > self.lines.len() {
```

Comment:
`Iterator::take` won't panic if there are less elements than `limit`, so this condition can be dropped—`take` already does what we need.

Reply:
Nice! I didn't even think about that. I didn't like that if statement anyway :smile: